### PR TITLE
Add new Go program to add an additional web auth token

### PR DESCRIPTION
**Pull request by alecthegeek**
This program is a standalone binary to make the installation of a new web services auth token easy

It will probably be used by third parties or installer engineers who need to automate adding new API based solutions and provide the token. No need to type values into the PaperCut MF/NG web admin interface.

### CODE REVIEW

**Comment by alecthegeek** [comment]
@tinhtruong Any chance you can review this please:-). You are hoping to use it.

File: PublicWebServicesAPI_AND_servercommandScripts/setUserProperties.php [hunk 1]
```diff
@@ -13,7 +13,7 @@
 
 function do_call($request) {
   
-  $url = "http://localhost:9191/rpc/api/xmlrpc";
+  $url = "http://172.18.48.1:9191/rpc/api/xmlrpc";
```
Comment (jamespercy) [inline PublicWebServicesAPI_AND_servercommandScripts/setUserProperties.php:16]: should we be hard coding an ip address?

Comment (alecthegeek) [reply PublicWebServicesAPI_AND_servercommandScripts/setUserProperties.php:16]: Oops  -- This content should not be here

File: PublicWebServicesAPI_AND_servercommandScripts/InstallNewWebAuthToken/InstallNewWebAuthToken.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+		update(auth)
+
+		return
+	}
+
+	var tokensAsArray []string
+
+	// Note: if the config key is not an array of strings this parse will fail
+	if err := json.Unmarshal(result, &tokensAsArray); err == nil {
```
Comment (jamespercy) [inline PublicWebServicesAPI_AND_servercommandScripts/InstallNewWebAuthToken/InstallNewWebAuthToken.go:125]: err needs to be handled

File: PublicWebServicesAPI_AND_servercommandScripts/InstallNewWebAuthToken/InstallNewWebAuthToken.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+			}
+		}
+
+		update(append(tokensAsArray, securityToken))
+		return
+	}
+
+	var tokensAsObject map[string]string
+
+	if err := json.Unmarshal(result, &tokensAsObject); err == nil {
```
Comment (jamespercy) [inline PublicWebServicesAPI_AND_servercommandScripts/InstallNewWebAuthToken/InstallNewWebAuthToken.go:141]: err needs to be handled

File: PublicWebServicesAPI_AND_servercommandScripts/InstallNewWebAuthToken/InstallNewWebAuthToken.go (hunk 2 above)
Comment (jamespercy) [inline PublicWebServicesAPI_AND_servercommandScripts/InstallNewWebAuthToken/InstallNewWebAuthToken.go:125]: the handling of err is a bit confusing here. maybe some logging or comment explaining might help. but not too bad I guess

**Comment by Joffcom** [comment]
What happens if the default path isn’t used for MF / NG on Windows which is something we see fairly often?

**Comment by alecthegeek** [comment]
I did think of that. But I figured it was an edge case.

Any tips on how to find the install location?

**Comment by Joffcom** [comment]
I think the only way to handle it would be to find the App Server Service and extract the path from that. I can’t think of any other method that would always work.

**Comment by Joffcom** [comment]
[This](https://godoc.org/golang.org/x/sys/windows/svc/mgr#Service.Config) has an option for BinaryPathName that could help.

**Comment by alecthegeek** [comment]
I've taken a halfway house for Windows install PATH.

Assume the env value PROGRAMFILES points to the correct install location

Happy to take PR for more sophisticated.

**Comment by alecthegeek** [comment]
merged and closed

**Pull request closed without merging**